File: Cplusplus/[1차] 다트 게임.cpp
//22.01.14 [1차] 다트 게임 by CJHofficial
#include <string>
#include <cmath>
#include <vector>

using namespace std;

int solution(string dartResult) {
    int answer = 0;
    vector<int> score;
    string NUM = "";
    for (auto s : dartResult) {
        if ((s >= '0') && (s <= '9')) {
            NUM += s;
        }
        else if (s == 'S') {
            score.push_back(stoi(NUM));
            NUM = "";
        }
        else if (s == 'D') {
            score.push_back(pow(stoi(NUM), 2));
            NUM = "";
        }
        else if (s == 'T') {
            score.push_back(pow(stoi(NUM), 3));
            NUM = "";
        }
        else if (s == '*') {
            score[score.size() - 1] = 2 * score[score.size() - 1];
            if (score.size() > 1) score[score.size() - 2] = 2 * score[score.size() - 2];
        }
        else if (s == '#') {
            score[score.size() - 1] = (-1) * score[score.size() - 1];
        }
    }
    for (int s : score) answer += s;
    return answer;
}

File: Python/level1/[1차] 다트 게임.py
import re
square = {'S':1,'D':2,'T':3}
def solution(dartResult):
    answer = []
    scores = re.compile('\d+[SDT][*#]*').findall(dartResult)
    for score in scores:
        number = re.search('\d+', score).group()
        word = re.search('[SDT]', score).group()
        option = re.search('[*#]', score)
        realScore = int(number) ** square[word]
        if option:
            if option.group() == '*':
                realScore = realScore * 2
                if len(answer) != 0:
                    answer[-1] = answer[-1] * 2
            elif option.group() == '#':
                realScore = -realScore
        answer.append(realScore)
    return sum(answer)
